test/jdk/java/nio/channels/Channels/TransferTo.java

```diff
@@ -161,6 +175,29 @@ public void testMoreThanTwoGB() throws IOException {
         assertEquals(Files.mismatch(sourceFile, targetFile), -1);
     }
 
+    /*
+     * Special test whether selectable channel based transfer throws blocking mode exception.
+     */
+    @Test
+    public void testIllegalBlockingMode() throws IOException {
+        // preparing empty file input and non-blocking selectable output
+        FileChannel fc = FileChannel.open(Files.createTempFile(null, null));
+        InputStream is1 = Channels.newInputStream(fc);
+        SelectableChannel sc = Pipe.open().sink().configureBlocking(false);
+        OutputStream os1 = Channels.newOutputStream((WritableByteChannel) sc);
+
+        // IllegalBlockingMode must be thrown when trying to perform a transfer
+        assertThrows(IllegalBlockingModeException.class, () -> is1.transferTo(os1));
+
+        // preparing non-blocking selectable input and some arbitrary output
+        sc = Pipe.open().source().configureBlocking(false);
+        InputStream is2 = Channels.newInputStream((ReadableByteChannel) sc);
+        OutputStream os2 = new ByteArrayOutputStream();
+
+        // IllegalBlockingMode must be thrown when trying to perform a transfer
+        assertThrows(IllegalBlockingModeException.class, () -> is2.transferTo(os2));
```

Comment:
Thanks for adding this test. It looks like it leaks 5 file descriptors (the FileChannel, and the source/sink of two pipes). Can you update these to close the channels? The test current runs in othervm mode but if it is changed back to agentvm mode then leaving channels open could impact tests that are run later in the same VM.

Reply:
I sketched a possibly solution in https://github.com/openjdk/jdk/pull/5179/commits/26ed670be318669edf1ef9a7101c1136273f2b82, but looking at the PipeImpl internals I doubt that *all* file descriptors actually get closed. To me it feels like this being actually a design flaw of PipeImpl: If one side of a pipe was never actually claimed then it should not keep a descriptor open. So IMHO the final solution would be to file a PR for PipeImpl which either defers opening of the file descriptors until they are claimed (which I doubt would be easily possible), or which cleans up both sides if as soon as one side is closed and the other side never got claimed (which is my personal preference).

Reply:
> I sketched a possibly solution in [26ed670](https://github.com/openjdk/jdk/commit/26ed670be318669edf1ef9a7101c1136273f2b82), but looking at the PipeImpl internals I doubt that _all_ file descriptors actually get closed. To me it feels like this being actually a design flaw of PipeImpl: 

A Pipe creates a pair of channels. If you change to something the following then it will fix the issue:

```
Pipe pipe = Pipe.open();
try {
    :
} finally {
    pipe.source().close();
    pipe.sink().close();
}
```

Reply:
BTW: Would it possible to sync up your loom, it looks like hasn't been updated for some time.

Reply:
> A Pipe creates a pair of channels. If you change to something the following then it will fix the issue:

Sure, I know. Did that in https://github.com/openjdk/jdk/pull/5179/commits/26ed670be318669edf1ef9a7101c1136273f2b82. 

What I actually meant is: This is not apparent to the average Javadoc reader, so I propose to change PipeImpl to not run into this situation instead of asking all calling code to care for that.

Reply:
>BTW: Would it possible to sync up your loom, it looks like hasn't been updated for some time.

Uhm... syncing *what*?

Reply:
> Uhm... syncing _what_?

There was a typo, sorry. I meant your repo as the last commit from the main line is end Oct.

Reply:
> What I actually meant is: This is not apparent to the average Javadoc reader, so I propose to change PipeImpl to not run into this situation instead of asking all calling code to care for that.

A Pipe is two channels, one for each end. A typical usage will be for one thread to write to the sink, and another to read from the source. In the test you are only using one end, ignoring the other, so probably unusual. So while it would be possible to retrofit Pipe to implement AutoCloseable, it probably wouldn't fit many usages.

Reply:
Agreed. But back to original topic: Is the PR now in an acceptable shape?

Reply:
> Agreed. But back to original topic: Is the PR now in an acceptable shape?

I don't have any other comments but I think Lance does make a good point about the temp files.

Reply:
Done. :-)